pkg/util/quotapool/int_rate_impl_test.go

```diff
@@ -0,0 +1,190 @@
+// released.
+func (rl *RateLimiter) WaitN(ctx context.Context, n int64) error {
+	if n == 0 {
+		// Special case 0 acquisition.
+		return nil
+	}
+	r := rl.newRateRequest(n)
+	defer rl.putRateRequest(r)
+	if err := rl.qp.Acquire(ctx, r); err != nil {
+		return err
+	}
+	return r.err
+}
+
+// rateBucket is the implementation of Resource which remains in the quotapool
+// for a RateLimiter.
+type rateBucket struct {
+	p           *RateLimiter
+	cur         float64
+	lastUpdated time.Time
+}
+
+var _ Resource = (*rateBucket)(nil)
+
+func (i *rateBucket) Merge(other Resource) {
+	o := other.(*rateAlloc)
+	i.cur += float64(o.alloc)
+	o.rl.putRateAlloc(o)
+
+	if i.cur > float64(i.p.burst) {
+		i.cur = float64(i.p.burst)
+	}
+}
+
+// RateAlloc is an allocated quantity of quota which can be released back into
+// the token-bucket RateLimiter.
+type RateAlloc struct {
+	alloc int64
+	rl    *RateLimiter
+}
+
+// Return returns the RateAlloc to the RateLimiter. It is not safe to call any
+// methods on the RateAlloc after this call.
+func (ra *RateAlloc) Return() {
+	ra.rl.qp.Add((*rateAlloc)(ra))
+}
+
+// Consume destroys the RateAlloc. It is not safe to call any methods on the
+// RateAlloc after this call.
+func (ra *RateAlloc) Consume() {
+	ra.rl.putRateAlloc((*rateAlloc)(ra))
+}
+
+// rateAlloc is the internal implementation of Resource used by the RateLimiter.
+type rateAlloc RateAlloc
+
+var _ Resource = (*rateAlloc)(nil)
+
+func (i *rateAlloc) Merge(other Resource) {
+	panic("merge not implemented on rateAlloc")
+}
+
+type rateRequest struct {
+	want int64
+	err  error
+}
+
+var rateRequestSyncPool = sync.Pool{
+	New: func() interface{} { return new(rateRequest) },
+}
+
+// newRateRequest allocates a rateRequest from the sync.Pool.
+// It should be returned with putRateRequest.
+func (rl *RateLimiter) newRateRequest(v int64) *rateRequest {
+	r := rateRequestSyncPool.Get().(*rateRequest)
+	*r = rateRequest{want: v}
+	return r
+}
+
+func (rl *RateLimiter) putRateRequest(r *rateRequest) {
+	*r = rateRequest{}
+	rateRequestSyncPool.Put(r)
+}
+
+func (i *rateRequest) Acquire(
+	ctx context.Context, res Resource,
+) (fulfilled bool, unused Resource, tryAgainAfter time.Duration) {
+	r := res.(*rateBucket)
+	now := r.p.qp.timeSource.Now()
+	// Deal with the case where the allocation is larger than the burst size.
+	if i.want > r.p.burst {
+		i.err = ErrNotEnoughQuota
+		return true, res, 0
+	}
+
+	// TODO(ajwerner): Consider instituting a minimum update frequency to avoid
+	// spinning too fast on timers for tons of tiny allocations at a fast rate.
+	if since := now.Sub(r.lastUpdated); since > 0 {
+		r.cur += float64(r.p.rateLimit) * since.Seconds()
+		if r.cur > float64(r.p.burst) {
+			r.cur = float64(r.p.burst)
+		}
+		r.lastUpdated = now
+	}
+	if i.want > int64(r.cur) {
+		// Compute the time it will take for r.cur to get to the needed capacity.
+		delta := float64(i.want) - r.cur
```

Comment:
Can there be cases in which `i.want > int64(r.cur)` but `float64(i.want) < r.cur`?  The conversions here just make me anxious. Can this become

```go
if delta := float64(i.want) - r.cur; delta > 0 {
...
}
```